// @astrojs/underscore-redirects/dist/astro.js
import { posix } from "node:path";
import { Redirects } from "./redirects.js";
const pathJoin = posix.join;
function getRedirectStatus(route) {
  if (typeof route.redirect === "object") {
    return route.redirect.status;
  }
  return 301;
}
function createRedirectsFromAstroRoutes({
  config,
  routeToDynamicTargetMap,
  dir,
  buildOutput,
  assets
}) {
  const base = config.base && config.base !== "/" ? config.base.endsWith("/") ? config.base.slice(0, -1) : config.base : "";
  const redirects = new Redirects();
  for (const [route, dynamicTarget = ""] of routeToDynamicTargetMap) {
    const distURL = assets.get(route.pattern);
    if (route.pathname) {
      if (route.redirect) {
        redirects.add({
          dynamic: false,
          input: `${base}${route.pathname}`,
          target: typeof route.redirect === "object" ? route.redirect.destination : route.redirect,
          status: getRedirectStatus(route),
          weight: 2
        });
        continue;
      }
      if (buildOutput === "static") {
        continue;
      }
      if (distURL) {
        redirects.add({
          dynamic: false,
          input: `${base}${route.pathname}`,
          target: prependForwardSlash(distURL.toString().replace(dir.toString(), "")),
          status: 200,
          weight: 2
        });
      } else {
        redirects.add({
          dynamic: false,
          input: `${base}${route.pathname}`,
          target: dynamicTarget,
          status: 200,
          weight: 2
        });
        if (route.pattern === "/404") {
          redirects.add({
            dynamic: true,
            input: "/*",
            target: dynamicTarget,
            status: 404,
            weight: 0
          });
        }
      }
    } else {
      const pattern = generateDynamicPattern(route);
      if (distURL) {
        const targetRoute = route.redirectRoute ?? route;
        let target = generateDynamicPattern(targetRoute);
        if (config.build.format === "directory") {
          target = pathJoin(target, "index.html");
        } else {
          target += ".html";
        }
        redirects.add({
          dynamic: true,
          input: `${base}${pattern}`,
          target,
          status: route.type === "redirect" ? 301 : 200,
          weight: 1
        });
      } else {
        redirects.add({
          dynamic: true,
          input: `${base}${pattern}`,
          target: dynamicTarget,
          status: 200,
          weight: 1
        });
      }
    }
  }
  return redirects;
}
function generateDynamicPattern(route) {
  return "/" + route.segments.map(([part]) => {
    if (part.dynamic) {
      if (part.spread) {
        return "*";
      } else {
        return ":" + part.content;
      }
    } else {
      return part.content;
    }
  }).join("/");
}
function prependForwardSlash(str) {
  return str.startsWith("/") ? str : "/" + str;
}
export {
  createRedirectsFromAstroRoutes
};
